airbyte-cdk/bulk/core/load/src/main/kotlin/io/airbyte/cdk/load/component/TableSchemaEvolutionClient.kt

```diff
@@ -57,5 +57,156 @@ interface TableSchemaEvolutionClient {
         stream: DestinationStream,
         tableName: TableName,
         columnNameMapping: ColumnNameMapping,
-    )
+    ) {
+        val (actualSchema, actualAdditionalInfo) = discoverSchema(tableName)
+        val (expectedSchema, expectedAdditionalInfo) = computeSchema(stream, columnNameMapping)
+        val schemaDiff = actualSchema.diff(expectedSchema)
+        val additionalInfoDiff = diff(actualAdditionalInfo, expectedAdditionalInfo)
+        applySchemaDiff(
+            stream,
+            columnNameMapping,
+            tableName,
+            expectedSchema,
+            expectedAdditionalInfo,
+            schemaDiff,
+            additionalInfoDiff
+        )
+    }
+
+    /**
+     * Query the destination and discover the schema of an existing table. If this method includes
+     * the `_airbyte_*` columns, then [computeSchema] MUST also include those columns.
+     *
+     * [AdditionalSchemaInfo] should be used for anything not represented in the [TableSchema]
+     * object. For example, you may want to configure a partitioning/clustering key on the table,
+     * based on the sync mode - this can be provided via AdditionalSchemaInfo.
+     *
+     * Most destinations will likely use [Unit] for their additional info.
+     */
+    suspend fun discoverSchema(tableName: TableName): Pair<TableSchema, AdditionalSchemaInfo> {
+        throw NotImplementedError()
+    }
+
+    /**
+     * Compute the schema that we _expect_ the table to have, given the [stream]. This should _not_
+     * query the destination in any way.
+     *
+     * See [discoverSchema] for an explanation of [AdditionalSchemaInfo].
+     */
+    fun computeSchema(
+        stream: DestinationStream,
+        columnNameMapping: ColumnNameMapping
+    ): Pair<TableSchema, AdditionalSchemaInfo> {
+        throw NotImplementedError()
+    }
+
+    /**
+     * See [discoverSchema] for an explanation of [AdditionalSchemaInfo].
+     *
+     * This function computes a diff between two "additional info" structs.
+     *
+     * If your destination uses [Unit] as its additional info, you should use this implementation:
+     * ```kotlin
+     * fun diff(actualSchemaInfo: Unit, expectedSchemaInfo: Unit) = Unit
+     * ```
+     */
+    fun diff(
+        actualSchemaInfo: AdditionalSchemaInfo,
+        expectedSchemaInfo: AdditionalSchemaInfo
+    ): AdditionalSchemaInfoDiff {
+        throw NotImplementedError()
+    }
+
+    /**
+     * Execute the diff against the destination. After this method completes, a call to
+     * [discoverSchema] should return an identical schema as [computeSchema].
```

Comment:
this isn't striiiictly true - s3-data-lake actually tries to compute a "safe supertype" between the old+new types https://github.com/airbytehq/airbyte/blob/master/airbyte-cdk/bulk/toolkits/load-iceberg-parquet/src/main/kotlin/io/airbyte/cdk/load/toolkits/iceberg/parquet/IcebergTableSynchronizer.kt#L18

and snowflake's implementation never alters a column to become `NOT NULL` - https://github.com/airbytehq/airbyte/pull/69117/files?w=1#diff-d7a1a3320f7a7de452c17aadce76ab7d45be190647b1a0ad33c93fd41500a97cR434

... though I could be convinced that both of those behaviors are things we want to not do